File: 2022CompetitionBot/src/main/cpp/subsystems/DrivetrainSub.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "subsystems/DrivetrainSub.h"
#include <iostream>

constexpr double kShiftUpSpeed = 1.8; //meters per second
constexpr double kShiftDownSpeed = 1.4; 

constexpr double kEncoderRotationsToMetersLowGear = 5.0/204.44; //Find these actual values
constexpr double kEncoderRotationsToMetersHighGear = 5.0/129.5;
constexpr bool kGyroReversed = false;

DrivetrainSub::DrivetrainSub() {
  init(); 
}

void DrivetrainSub::init() { //Reset all hardware to a safe state
    m_leftMotor1.SetInverted(true); //Find which motor is acutally reversed and you can remove the false SetInverted functions
    m_leftMotor2.SetInverted(true);
    m_leftMotor3.SetInverted(true);

    m_rightMotor1.SetInverted(false);
    m_rightMotor2.SetInverted(false);
    m_rightMotor3.SetInverted(false);

    zeroDrivetrainEncoders();
    tankDrive(0., 0.);
    shiftDown();
    setIsAutoShift(true);

    m_gyro.Reset();
}

void DrivetrainSub::Periodic() {}

//Drive functions
void  DrivetrainSub::tankDrive(double lPower, double rPower) {
    m_drive.TankDrive(lPower, rPower);
}

void DrivetrainSub::arcadeDrive(double drivePwr, double rotatePwr) {
    m_drive.TankDrive(drivePwr+rotatePwr, drivePwr-rotatePwr);
}

void  DrivetrainSub::shiftUp() { //Gear shift up
    m_shifter.Set(1);
}

void  DrivetrainSub::shiftDown() { // Gear shift down
    m_shifter.Set(0);
}

void DrivetrainSub::autoShift() {
    if (!m_isAutoShift){
        return;
    }
    //get speed from robot
    double averageWheelSpeed = (getLeftVelocity() + getRightVelocity()) / 2.;

    if(fabs(averageWheelSpeed) > kShiftUpSpeed) {
        shiftUp();
    }
    else if(fabs(averageWheelSpeed) < kShiftDownSpeed) {
        shiftDown();
    }
}

bool DrivetrainSub::isShiftedInHighGear() {
    return m_shifter.Get();
}

void DrivetrainSub::zeroDrivetrainEncoders() {
    m_leftMotorEncoder.SetPosition(0.);
    m_rightMotorEncoder.SetPosition(0.);
}

void DrivetrainSub::zeroHeading() {
    m_gyro.Reset();
}

double DrivetrainSub::getLeftEncoderRaw() { //Returns rotations (1 full motor rotation = 1.)
    return m_leftMotorEncoder.GetPosition();
}

double DrivetrainSub::getRightEncoderRaw() { 
    return m_rightMotorEncoder.GetPosition();
}

double DrivetrainSub::getLeftEncoderDistanceM() {
    return getLeftEncoderRaw() * getEncoderRotationsToMeterFactor();
}
double DrivetrainSub::getRightEncoderDistanceM() {
    return getRightEncoderRaw() * getEncoderRotationsToMeterFactor();
}

double DrivetrainSub::getEncoderDistanceM() {
    return (getLeftEncoderDistanceM()+getRightEncoderDistanceM())/2;
}

double DrivetrainSub::getEncoderRotationsToMeterFactor() {
    return (isShiftedInHighGear()) ? kEncoderRotationsToMetersHighGear : kEncoderRotationsToMetersLowGear;
}

double DrivetrainSub::getLeftVelocity() {
    return m_leftMotorEncoder.GetVelocity() * getEncoderRotationsToMeterFactor() / 60.; //In meters per second
}

double DrivetrainSub::getRightVelocity() {
    return m_rightMotorEncoder.GetVelocity() * getEncoderRotationsToMeterFactor() / 60.; //In meters per second
}

double DrivetrainSub::getVelocity() {
    return (getLeftVelocity()+getRightVelocity())/2;
}

double DrivetrainSub::getHeading() {
    return m_gyro.GetAngle() * (kGyroReversed ? -1. : 1.);
}

double DrivetrainSub::getTurnRate () {
    return m_gyro.GetRate() * (kGyroReversed ? -1. : 1.);
}

void DrivetrainSub::setIsAutoShift(bool autoShiftActive) {
    m_isAutoShift = autoShiftActive;
    return;
}

File: Training/4917RobotTemplate - Eli and Beth v2.0/src/main/java/frc/robot/subsystems/DrivetrainSub.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.subsystems;

import edu.wpi.first.wpilibj2.command.SubsystemBase;
import frc.robot.Constants;
import frc.robot.subsystems.DrivetrainSub;
import com.revrobotics.CANSparkMax;
import com.revrobotics.CANSparkMaxLowLevel;

public class DrivetrainSub extends SubsystemBase {
  private final CANSparkMax m_leftMotor1 = new CANSparkMax(Constants.CanIds.kLeftMotor1, CANSparkMaxLowLevel.MotorType.kBrushless);
  private final CANSparkMax m_leftMotor2 = new CANSparkMax(Constants.CanIds.kLeftMotor2, CANSparkMaxLowLevel.MotorType.kBrushless);
  private final CANSparkMax m_leftMotor3 = new CANSparkMax(Constants.CanIds.kLeftMotor3, CANSparkMaxLowLevel.MotorType.kBrushless);
  
  private final CANSparkMax m_rightMotor1 = new CANSparkMax(Constants.CanIds.kRightMotor1, CANSparkMaxLowLevel.MotorType.kBrushless);
  private final CANSparkMax m_rightMotor2 = new CANSparkMax(Constants.CanIds.kRightMotor2, CANSparkMaxLowLevel.MotorType.kBrushless);
  private final CANSparkMax m_rightMotor3 = new CANSparkMax(Constants.CanIds.kRightMotor3, CANSparkMaxLowLevel.MotorType.kBrushless);

  /** Creates a new DrivetrainSub. */
  public DrivetrainSub() {}

  @Override
  public void periodic() {
    // This method will be called once per scheduler run
  }
  public void tankDrive(double leftPower, double rightPower) {
    m_leftMotor1.set(leftPower);
    m_leftMotor2.set(leftPower);
    m_leftMotor3.set(leftPower);
    m_rightMotor1.set(rightPower);
    m_rightMotor2.set(rightPower);
    m_rightMotor3.set(rightPower);
  }

public void arcadeDrive(double d, double e) {
}
}
